go/vt/vtgate/planbuilder/operators/apply_join.go

```diff
@@ -159,57 +182,113 @@ func (a *ApplyJoin) pushColRight(ctx *plancontext.PlanningContext, e *sqlparser.
 }
 
 func (a *ApplyJoin) GetColumns() ([]sqlparser.Expr, error) {
-	return a.ColumnsAST, nil
+	columns := slices2.Map(a.ColumnsAST, jcToExpr)
+	return columns, nil
 }
 
-func (a *ApplyJoin) AddColumn(ctx *plancontext.PlanningContext, expr *sqlparser.AliasedExpr, reuseCol bool) (ops.Operator, int, error) {
-	if offset, found := canReuseColumn(ctx, reuseCol, a.ColumnsAST, expr.Expr); found {
-		return a, offset, nil
-	}
+func jcToExpr(c JoinColumn) sqlparser.Expr { return c.Original }
 
+func (a *ApplyJoin) getJoinColumnFor(ctx *plancontext.PlanningContext, e sqlparser.Expr) (col JoinColumn, err error) {
+	col.Original = e
 	lhs := TableID(a.LHS)
 	rhs := TableID(a.RHS)
 	both := lhs.Merge(rhs)
-	deps := ctx.SemTable.RecursiveDeps(expr.Expr)
+	deps := ctx.SemTable.RecursiveDeps(e)
 
-	// if we get here, it's a new expression we are dealing with.
-	// We need to decide if we can push it all on either side,
-	// or if we have to break the expression into left and right parts
 	switch {
 	case deps.IsSolvedBy(lhs):
-		offset, err := a.pushColLeft(ctx, expr)
-		if err != nil {
-			return nil, 0, err
-		}
-		a.Columns = append(a.Columns, -offset-1)
+		col.LHSExprs = []sqlparser.Expr{e}
 	case deps.IsSolvedBy(rhs):
-		offset, err := a.pushColRight(ctx, expr)
-		if err != nil {
-			return nil, 0, err
-		}
-		a.Columns = append(a.Columns, offset+1)
+		col.RHSExpr = e
 	case deps.IsSolvedBy(both):
-		bvNames, lhsExprs, rhsExpr, err := BreakExpressionInLHSandRHS(ctx, expr.Expr, lhs)
+		col, err = BreakExpressionInLHSandRHS(ctx, e, TableID(a.LHS))
 		if err != nil {
-			return nil, 0, err
+			return JoinColumn{}, err
 		}
-		for i, lhsExpr := range lhsExprs {
+	default:
+		return JoinColumn{}, vterrors.VT13002(sqlparser.String(e))
+	}
+
+	return
+}
+
+func (a *ApplyJoin) AddColumn(ctx *plancontext.PlanningContext, expr *sqlparser.AliasedExpr) (ops.Operator, int, error) {
+	if offset, found := canReuseColumn(ctx, a.ColumnsAST, expr.Expr, jcToExpr); found {
+		return a, offset, nil
+	}
+	col, err := a.getJoinColumnFor(ctx, expr.Expr)
+	if err != nil {
+		return nil, 0, err
+	}
+	a.ColumnsAST = append(a.ColumnsAST, col)
+	return a, len(a.ColumnsAST) - 1, nil
+}
+
+func (a *ApplyJoin) planOffsets(ctx *plancontext.PlanningContext) error {
+	for _, col := range a.ColumnsAST {
+		// Read the type description for JoinColumn to understand the following code
+		for i, lhsExpr := range col.LHSExprs {
 			offset, err := a.pushColLeft(ctx, aeWrap(lhsExpr))
 			if err != nil {
-				return nil, 0, err
+				return err
+			}
+			if col.RHSExpr == nil {
+				// if we don't have a RHS expr, it means that this is a pure LHS expression
+				a.addOffset(-offset - 1)
```

Comment:
it's not exactly clear why we are doing `-offset` here. I am guessing it's because offsets for `LHS` are going to be negative, and we want to turn it into a positive here, but I am not really sure.

Reply:
Makes sense. Now I remember :)